src/doc/trpl/choosing-guarantees.md

```diff
@@ -0,0 +1,417 @@
+
+## `&T` and `&mut T` 
+
+These are immutable and mutable references respectively. They follow the "read-write lock" pattern,
+such that one may either have only one mutable reference to some data, or any number of immutable
+ones, but not both. This guarantee is enforced at compile time, and has no visible cost at runtime.
+In most cases such pointers suffice for sharing cheap references between sections of code.
+
+These pointers cannot be copied in such a way that they outlive the lifetime associated with them.
+
+## `*const T` and `*mut T`
+
+These are C-like raw pointers with no lifetime or ownership attached to them. They just point to
+some location in memory with no other restrictions. The only guarantee that these provide is that
+they cannot be dereferenced except in code marked `unsafe`.
+
+These are useful when building safe, low cost abstractions like `Vec<T>`, but should be avoided in
+safe code.
+
+## `Rc<T>`
+
+This is the first wrapper we will cover that has a runtime cost.
+
+
+[`Rc<T>`][rc] is a reference counted pointer. In other words, this lets us have multiple "owning"
+pointers to the same data, and the data will be freed (destructors will be run) when all pointers
+are out of scope.
+
+Internally, it contains a shared "reference count", which is incremented each time the `Rc` is
+cloned, and decremented each time one of the `Rc`s goes out of scope. The main responsibility of
+`Rc<T>` is to ensure that destructors are called for shared data.
+
+The internal data here is immutable, and if a cycle of references is created, the data will be
+leaked. If we want data that doesn't leak when there are cycles, we need a garbage collector.
+
+// Should I link to rust-gc and rajan-bacon-cc here?
+
+
+#### Guarantees
+
+The main guarantee provided here is that the data will not be destroyed until all references to it
+are out of scope.
+
+This should be used when you wish to dynamically allocate and share some data (read-only) between
+various portions of your program, where it is not certain which portion will finish using the
+pointer last. It's a viable alternative to `&T` when `&T` is either impossible to statically check
+for correctness, or creates extremely unergonomic code where the programmer does not wish to spend
+the development cost of working with.
+
+This pointer is _not_ thread safe, and Rust will not let it be sent or shared with other threads.
+This lets one avoid the cost of atomics in situations where they are unnecessary.
+
+
+There is a sister smart pointer to this one, `Weak<T>`. This is a non-owning, but also non-borrowed,
+smart pointer. It is also similar to `&T`, but it is not restricted in lifetime &mdash; a `Weak<T>`
+can be held on to forever. However, it is possible that an attempt to access the inner data may fail
+and return `None`, since this can outlive the owned `Rc`s. This is useful for when one wants cyclic
+data structures and other things.
```

Comment:
> This is useful for when one wants cyclic data structures and other things.

Maybe just "This is useful for cyclic data structures and other things."? And also expand on "other things" :)